### PR TITLE
Jrpcfs shutdown fix

### PR DESCRIPTION
**Pull request by balajisrao**

### CODE REVIEW

File: jrpcfs/config.go [hunk 1]
```diff
@@ -33,6 +35,12 @@ type globalsStruct struct {
 
 	// Map used to store volumes already mounted for bimodal support
 	bimodalMountMap map[string]fs.MountHandle
+
+	// Connection list and listner list to close during shutdown:
```
Comment (edmc-ss) [inline jrpcfs/config.go]: ^listner^listener

File: jrpcfs/config.go [hunk 2]
```diff
@@ -33,6 +35,12 @@ type globalsStruct struct {
...
+	connLock      sync.Mutex
+	connections   *list.List
+	connWaitGroup sync.WaitGroup
+	listners      []net.Listener
```
Comment (edmc-ss) [inline jrpcfs/config.go]: !!

File: jrpcfs/config.go [hunk 3]
```diff
@@ -115,6 +123,8 @@ func Up(confMap conf.ConfMap) (err error) {
 		}
 	}
 
+	globals.connections = list.New()
```
Comment (edmc-ss) [inline jrpcfs/config.go]: So you are just using one connections list to cover both types of connections?

Comment (edmc-ss) [reply jrpcfs/config.go]: And don't you also need to globals.listners = make([]net.Listener)?

File: jrpcfs/filesystem.go [hunk 4]
```diff
@@ -20,7 +20,6 @@ import (
 
 // RPC server handle
 var srv *rpc.Server
```
Comment (edmc-ss) [inline jrpcfs/filesystem.go:22]: I think these three should now be local and in .globals... not here, right?

File: jrpcfs/io.go [hunk 5]
```diff
@@ -41,18 +46,27 @@ func ioServerLoop() {
...
 
-		go ioHandle(conn)
+		globals.connWaitGroup.Add(1)
+
+		globals.connLock.Lock()
+		globals.connections.PushBack(conn)
+		globals.connLock.Unlock()
+
+		go func() {
+			ioHandle(conn)
```
Comment (edmc-ss) [inline jrpcfs/io.go]: Don't we need a:
```
globals.connLock.Lock()
globals.connections.Remove(conn)
globals.connLock.Unlock()
```
here?

File: jrpcfs/filesystem.go [hunk 6]
```diff
@@ -53,17 +57,26 @@ func jrpcServerLoop() {
...
 
-		go srv.ServeCodec(jsonrpc.NewServerCodec(conn))
+		globals.connWaitGroup.Add(1)
+
+		globals.connLock.Lock()
+		globals.connections.PushBack(conn)
+		globals.connLock.Unlock()
+
+		go func() {
+			srv.ServeCodec(jsonrpc.NewServerCodec(conn))
```
Comment (edmc-ss) [inline jrpcfs/filesystem.go]: Don't we need a:
```
globals.connLock.Lock()
globals.connections.Remove(conn)
globals.connLock.Unlock()
```
here?

**Review by edmc-ss — CHANGES_REQUESTED** [review]
Looks pretty close...
My only technical suggestion was to ensure you remove conn's from the connections list as their goroutines exit...

File: jrpcfs/config.go [hunk 7]
```diff
@@ -299,5 +309,16 @@ func Down() (err error) {
 	err = nil
 	jsonRpcServerDown()
 	ioServerDown()
+
+	globals.Lock()
+	for _, listner := range globals.listners {
+		listner.Close()
+	}
+
+	for elm := globals.connections.Front(); elm != nil; elm = elm.Next() {
```
Comment (edmc-ss) [inline jrpcfs/config.go]: Ah... and per my earlier comment about this connections list being updated as each goroutine exits... this looping would need to be careful to lock/unlock around accesses...

The way you've done it... you are just sending close's... to a lock/unlock around the loop is fine.

File: jrpcfs/config.go [hunk 8]
```diff
@@ -299,5 +309,16 @@ func Down() (err error) {
...
+		conn := elm.Value.(net.Conn)
+		conn.Close()
+	}
+	globals.Unlock()
```
Comment (edmc-ss) [inline jrpcfs/config.go]: And, finally, I expected to see a globals. connWaitGroup.Wait() call here, right?

**Review by edmc-ss — CHANGES_REQUESTED** [review]
Ok - I had another technical set of comments... basically I was looking for the Wait() call to await both the Listeners and all the child goroutines' completions.

File: jrpcfs/config.go (hunk 1 above)
Comment (balajisrao) [reply jrpcfs/config.go]: Done!

File: jrpcfs/config.go (hunk 2 above)
Comment (balajisrao) [reply jrpcfs/config.go]: Done

File: jrpcfs/io.go (hunk 5 above)
Comment (balajisrao) [reply jrpcfs/io.go]: YES. fixed it.

File: jrpcfs/filesystem.go (hunk 6 above)
Comment (balajisrao) [reply jrpcfs/filesystem.go]: YES

File: jrpcfs/config.go (hunk 8 above)
Comment (balajisrao) [reply jrpcfs/config.go]: YES - fixed it.

File: jrpcfs/config.go (hunk 7 above)
Comment (balajisrao) [reply jrpcfs/config.go]: yes, that way the list is consistent till all close() calls are done and then the list becomes empty as each connections close.

**Review by edmc-ss — APPROVED** [review]
With the make([]net.Listener, 2) changed to make([]net.Listener, 0, 2), this looks perfect!

**Pull request merged**